Ensure that the S.N.Vector methods are marked as [Intrinsic] and that the attribute is respected by the runtime

tannergooding opened the pull request:
As noted here: https://github.com/dotnet/coreclr/pull/21518#issuecomment-448375887 and here: https://github.com/dotnet/corefx/issues/31425#issuecomment-448390800...

Many of the S.N.Vector methds were marked as `[Intrinsic]`. However, for the S.N.Vector methods, the runtime was not actually checking for the `[Intrinsic]` attribute and instead just looked at all methods in the types marked `[Intrinsic]`. There were indeed several methods not marked as `[Intrinsic]` that were being handled as intrinsics, there were also some methods marked as `[Intrinsic]` that were not appropriately hooked up in the runtime. This put us in a scenario where:
* You could not determine if a method could be treated as intrinsic or not by looking at the source code in CoreFX
* You could not locally disable intrinsic recognition of a method by removing the attribute (making investigations around https://github.com/dotnet/corefx/issues/31425 difficult)

This fixes up the runtime to start checking for the `[Intrinsic]` attribute. It also ensures that all methods we were already treating as intrinsic are appropriately marked as such.

tannergooding commented:
CoreFX side is here: https://github.com/dotnet/corefx/pull/34167

tannergooding commented:
CC. @CarolEidt, @fiigii

tannergooding commented:
I generated pmi diffs locally using these changes alongside https://github.com/dotnet/corefx/pull/34167 to validate there were no diffs.

tannergooding commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -5120,7 +5120,7 @@ MethodTableBuilder::InitNewMethodDesc(
 
     // Check for methods marked as [Intrinsic]
-    if (GetModule()->IsSystem())
+    if (GetModule()->IsSystem() || GetAssembly()->IsSIMDVectorAssembly())
```
This is the same check we do for the types: https://github.com/dotnet/coreclr/blob/7568519bf2a964139d39932534569bf5299347cb/src/vm/methodtablebuilder.cpp#L1527

tannergooding commented on src/jit/simd.cpp:
```diff
@@ -2349,11 +2349,14 @@ GenTree* Compiler::impSIMDIntrinsic(OPCODE                opcode,
-    if (!isSIMDClass(clsHnd))
+    // Exit early if we are either not in one of the SIMD types or if the method
+    // is not a JIT Intrinsic (which requires the [Intrinsic] attribute).
+    if (!isSIMDClass(clsHnd) || ((methodFlags & CORINFO_FLG_JIT_INTRINSIC) == 0))
```
When validating locally, I caught all the methods that we were implicitly treating as intrinsic (all the ones above that I added the `[Intrinsic]` attribute to) by adding an assert that `(methodFlags & CORINFO_FLG_JIT_INTRINSIC) != 0` on L2363 (so after the `getSIMDIntrinsicInfo` call and check that the intrinsic returned was valid).

I then generated PMI diffs for `-c` (corelib), `-f` (framework), `--benchmarks`, and `--tests`; which hit the assert for everything that was being implicitly treated as `[Intrinsic]`.

CarolEidt replied:
Did you get to zero diffs once everything was properly marked? It seems that you decided to make the change to `Dot`, so presumably that would lead to diffs?

tannergooding replied:
Yes, I got to zero-diffs once everything was properly marked.

With the change to `Dot` and `AndNot` (making them recognized), there are now diffs for those call-sites. I'll collect them again and post.

tannergooding replied:
@CarolEidt, I reran the diffs (`-c`, `-f`, `--benchmarks`, and `--tests`) using the current HEAD as base and this PR + the CoreFX side changes (https://github.com/dotnet/corefx/pull/34167) as the diff.

The only delta was in `S.P.Corelib` (due to `Vector.AndNot` now being treated as an intrinsic):
```
Found 1 files with textual diffs.
PMI Diffs for System.Private.CoreLib.dll for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -402 (-0.01% of base)
    diff is an improvement.
Top file improvements by size (bytes):
        -402 : System.Private.CoreLib.dasm (-0.01% of base)
1 total files with size differences (1 improved, 0 regressed), 0 unchanged.
Top method improvements by size (bytes):
        -402 (-41.40% of base) : System.Private.CoreLib.dasm - Vector`1:ConditionalSelect(struct,struct,struct):struct (4 methods)
Top method improvements by size (percentage):
        -402 (-41.40% of base) : System.Private.CoreLib.dasm - Vector`1:ConditionalSelect(struct,struct,struct):struct (4 methods)
1 total methods with size differences (1 improved, 0 regressed), 14260 unchanged.
Completed analysis in 3.13s
```

I can share the full delta if you're interested, but it basically just comes down to replacing the following large sequence of instructions:
```asm
       mov      rcx, 0xD1FFAB1E
       mov      edx, 252
       vextractf128 ymm8, ymm6, 1
       vextractf128 ymm9, ymm7, 1
       call     CORINFO_HELP_CLASSINIT_SHARED_DYNAMICCLASS
       vinsertf128 ymm7, ymm9, 1
       vinsertf128 ymm6, ymm8, 1
       mov      rax, 0xD1FFAB1E
       mov      rax, gword ptr [rax]
       vmovupd  ymm0, ymmword ptr[rax+8]
       vxorpd   ymm0, ymm7
       vandpd   ymm0, ymm6, ymm0
       vmovupd  ymm1, ymmword ptr[rdi]
       vmovupd  ymm2, ymmword ptr[rsi]
       vandpd   ymm1, ymm2
```

with:
```asm
vandnpd  ymm1, ymm2
```

tannergooding replied:
I also explicitly got diffs for `System.Numerics.Vectors.PerformanceTests`, which had no diffs.


and for `System.Numerics.Vectors.Tests`, which had a couple:

```
Found 1 files with textual diffs.
PMI Diffs for E:\Users\tagoo\Repos\corefx\artifacts\bin\System.Numerics.Vectors.Tests\netcoreapp-Release\System.Numerics.Vectors.Tests.dll for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -48 (-0.03% of base)
    diff is an improvement.
Top file improvements by size (bytes):
         -48 : System.Numerics.Vectors.Tests.dasm (-0.03% of base)
1 total files with size differences (1 improved, 0 regressed), 0 unchanged.
Top method improvements by size (bytes):
         -48 (-2.14% of base) : System.Numerics.Vectors.Tests.dasm - GenericVectorTests:TestBitwiseAndNot():this (4 methods)
Top method improvements by size (percentage):
         -48 (-2.14% of base) : System.Numerics.Vectors.Tests.dasm - GenericVectorTests:TestBitwiseAndNot():this (4 methods)
1 total methods with size differences (1 improved, 0 regressed), 980 unchanged.
Completed analysis in 0.45s
```

CarolEidt replied:
> I can share the full delta if you're interested

No need - just wanted to make sure someone had a look.

tannergooding commented:
This will make it much easier to reason about what is/is not intrinsic and will allow us to more easily iterate on https://github.com/dotnet/corefx/issues/31425

jkotas commented on src/System.Private.CoreLib/shared/System/Numerics/Vector.cs:
```diff
@@ -4182,7 +4196,6 @@ internal static unsafe Vector<T> Max(Vector<T> left, Vector<T> right)
         }
 
-        [Intrinsic]
         internal static T DotProduct(Vector<T> left, Vector<T> right)
```
`DotProduct` is recognized as intrinsic by .NET Native toolchain. 

The `[Intrinsic]` attributes are not exclusive to RyuJIT.

tannergooding replied:
> DotProduct is recognized as intrinsic by .NET Native toolchain.

Would it be better to just rename this and fixup `BitwiseAndNot` here, and .NET Native can be updated to recognize the `Dot` name (which is consistent with what we have named and handled everywhere else in CoreCLR)?

jkotas replied:
Sounds good to me.

tannergooding replied:
@jkotas or @MichalStrehovsky, if one of you could point me to the SIMD Intrinsic support in our .NET Native codebase (feel free to send me a mail offline), I'd be happy to fix this up there as well. I've previously made changes here for the Math intrinsics, but I don't seem to be able to find the SIMD intrinsic support.

MichalStrehovsky replied:
The mapping from the `DotProduct` name to the intrinsic is done in src\Nutc\inc\STS_HostDefinitions.h. This gets mapped to CORINFO_INTRINSIC_NumericsVectorTDot that is later handled in a couple files under src\Nutc\UTC. If it's just about renaming, editing the host definitions should be enough.

We might need to do this in a backwards compatible manner. I'm not clear on whether NUTC supports the OOB version of System.Numerics.Vectors that would still have the old name. Maybe @nattress would know.

tannergooding replied:
I've updated the e-mail thread with the appropriate shelveset and left the existing definitions in for back-compat.

tannergooding commented:
Looks like something is broken in the `System.Numerics.Tests.GenericVectorTests.BitwiseAndNot` tests... I'm going to take a quick look and see if it's something trivial.

tannergooding commented on src/jit/simd.cpp:
```diff
@@ -2879,6 +2882,18 @@ GenTree* Compiler::impSIMDIntrinsic(OPCODE                opcode,
             op1 = impSIMDPopStack(simdType, instMethod);
 
+#ifdef _TARGET_XARCH_
+            if (simdIntrinsicID == SIMDIntrinsicBitwiseAndNot)
```
This is the only new change and it ensures that the operands are passed in the expected order for XARCH, since it does `~op1 & op2`, while ARM and the software fallback do `op1 & ~op2`.

tannergooding commented:
Applied formatting patch.

tannergooding commented:
@jkotas, @CarolEidt. I believe I have addressed all of your feedback and the .NET Native side changes have been provided in the form of a shelveset.

If neither of you have any objections, I believe this and the CoreFX side changes (https://github.com/dotnet/corefx/pull/34167) are ready to be merged.
